### PR TITLE
Add "Edit on GitHub" links to each page / section. Fixes #2029

### LINKED ISSUE #2029: Links to "Edit this page on Github"
We potentially need more links back from the documentation site to the Github repo. 

I was thinking we might put an "Edit this page on Github" link in a (smaller) font (and maybe in italics) just under every level 2 header "==". Most (all?) level 2 headers are used as top-of-adoc file headers so it'd be an easy link back from the section to the specific asciidoc file that's contains the text. 

Might (will!) needs some toolchain magic perhaps to insert links automatically (or alternatively I could just go through and add them all manually, but that sounds like a bit of a nightmare to maintain long term.

Thoughts @lurch @nelliemckesson ?

### PR DESCRIPTION
**Pull request by lurch**
Quite a bit going on here... :sweat_smile: 
* I've added machinery to `.github/workflows/build.yml` for installing Python modules from `requirements.txt`
* I've added `requirements.txt` to install the [pyyaml](https://pyyaml.org/wiki/PyYAMLDocumentation) module
* ...so that we can make use of values in `_config.yml` from the python build-scripts
* ...which I've done in several places :slightly_smiling_face: 
* The way the "build" works has changed so that every `*.adoc` file now gets copied across to a subdirectory of `build/jekyll` (previously the files in `build/jekyll` just directly-included the adoc files straight from `documentation/asciidoc`)
* ...because the first header in each of the `build/jekyll/*.adoc` files now has an autogenerated "Edit this on GitHub" link added immediately underneath it :tada: 
* ...with the text in that link coming from `jekyll-assets/_includes/github_edit.adoc`

I'm going to mark this as draft because we want some @nelliemckesson styling applied to these new links, before this gets merged :slightly_smiling_face: 
Screenshot from 2021-08-13 16-34-35

### CODE REVIEW

**Comment by aallan** [comment]
@nelliemckesson So I think we need to make the font of "Edit this on Github" smaller, maybe italic(/), and close the line up to hug the Level 2 header pretty closely underneath (so remove all that whitespace). Any thoughts on how it should look @jackbenwillis?

File: scripts/create_jekyll_ninjabuild.py [hunk 1]
```diff
@@ -32,13 +33,21 @@ def scan_adoc(adoc_filename, apparent_filename):
...
+    config_yaml = sys.argv[2]
+    input_dir = sys.argv[3]
     if not os.path.exists(input_dir):
         raise Exception("Error: {} doesn't exist".format(input_dir))
-    output_dir = sys.argv[3]
-    output_ninjabuild = sys.argv[4]
+    output_dir = sys.argv[4]
+    output_ninjabuild = sys.argv[5]
 
-    category_pages = set([('index.adoc', 'Raspberry Pi Documentation'), ('404.adoc', 'Raspberry Pi Documentation')])
```
Comment (lurch) [inline scripts/create_jekyll_ninjabuild.py:41]: I'm so glad that I managed to get rid of the "niggle" of having these hardcoded page-title strings embedded inside these python build-scripts :slightly_smiling_face:

File: scripts/create_build_adoc_include.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    with open(config_yaml) as config_fh:
+        site_config = yaml.safe_load(config_fh)
+    with open(src_adoc) as in_fh:
+        new_contents = ''
+        seen_header = False
+        for line in in_fh.readlines():
+            if line.startswith('== '):
+                if not seen_header:
+                    seen_header = True
+                    line += "Edit this {}[on GitHub]\n\n".format(os.path.join(site_config['githuburl'], 'blob', site_config['githubbranch_edit'], src_adoc))
```
Comment (lurch) [inline scripts/create_build_adoc_include.py:21]: So there's a small niggle that this "Edit this on GitHub" appears in both `scripts/create_build_adoc.py` and `scripts/create_build_adoc_include.py`, but I'm not bothered enough about it to fix it yet :wink:

**Comment by jackbenwillis** [comment]
> @nelliemckesson So I think we need to make the font of "Edit this on Github" smaller, maybe italic(?), and close the line up to hug the Level 2 header pretty closely underneath (so remove all that whitespace).
> 
> Any thoughts on how it should look @jackbenwillis?

Could we do what wikipedia do?
image
Align to the section heading, square brackets?, and as Alastair mentioned – a tad smaller than the body copy.

**Comment by andrum99** [comment]
> > @nelliemckesson So I think we need to make the font of "Edit this on Github" smaller, maybe italic(?), and close the line up to hug the Level 2 header pretty closely underneath (so remove all that whitespace).
> > Any thoughts on how it should look @jackbenwillis?
> 
> Could we do what wikipedia do?
> image
> Align to the section heading, square brackets?, and as Alastair mentioned – a tad smaller than the body copy.

I don't like Wikipedia's way of doing it - to me it makes the site look like it is still a hard hat area, particularly the square brackets. My preference is for the original proposal, except remove the word `this` since it is redundant. So that becomes `Edit on Github`. Also, it looks cleaner if the link covers the whole line of text.

**Comment by aallan** [comment]
I'm inclined to see what @nelliemckesson can do to hug the text up underneath the Level 2 headers. The Wikipedia way of doing things is very much associated with the the MediaWiki stack which could confuse a lot of people.

**Comment by aallan** [comment]
But I quite like the current phraseology, and I think we explicitly want the link not to run the whole text.

File: scripts/create_build_adoc_include.py (hunk 2 above)
Comment (lurch) [reply scripts/create_build_adoc_include.py]: Now fixed! :laughing:

**Comment by lurch** [comment]
It sounds like there's potential for the text around this edit-link to be tweaked, so I've now moved it into it's own template file (see `jekyll-assets/_includes/github_edit.adoc`) so that `scripts/create_build_adoc.py` and `scripts/create_build_adoc_include.py` won't ever become "out of sync" :slightly_smiling_face: 

(Strictly speaking this isn't a **Jekyll** include, as it only gets used by the python build-scripts and not by jekyll, but this seemed like the best place to put this new template file.)

**Comment by lurch** [comment]
After this has been merged, it might be worth mentioning these "Edit" links in `CONTRIBUTING.md` ?

**Comment by aallan** [comment]
> After this has been merged, it might be worth mentioning these "Edit" links in `CONTRIBUTING.md` ?

Yup. Totally.

**Comment by nelliemckesson** [comment]
Just submitted PR #2092 into this branch that adds styling to make the links look like this:

Screen Shot 2021-08-17 at 8 13 02 PM

Edits are easy, so don't be shy :)

**Comment by aallan** [comment]
> Just submitted PR #2092 into this branch that adds styling…

I like this. I'm going to go ahead and merge #2092 to give us the functionality and get those toolchain changes into production. If we feel there needs to be stylistic changes after we see this in action we can raise a new PR.

**Comment by aallan** [comment]
There was a merge conflict and I went with `margin-bottom: 0px` instead of `margin-bottom: 15px` for the H2 element. Let's see if that was right.

**Review by aallan — APPROVED** [review]

**Comment by aallan** [comment]
Looks okay on local, @nelliemckesson was that the right `margin-bottom` ..?

**Comment by lurch** [comment]
LGTM for the most part, but there's a few places (that I've spotted) where it looks a bit strange...

On http://127.0.0.1:4000/documentation/accessories/sense-hat.html
Screenshot from 2021-08-18 11-52-21

On http://127.0.0.1:4000/documentation/accessories/tv-hat.html
Screenshot from 2021-08-18 11-55-02

Also, I wonder if it might look better if the "Edit this on GitHub" had some left-padding so that it was indented slightly? :shrug:

**Comment by aallan** [comment]
> LGTM for the most part, but there's a few places (that I've spotted) where it looks a bit strange...

Maybe that `margin-bottom` should have been `15px` not `0px` ..?

**Comment by aallan** [comment]
No, playing around with it locally, definitely should be `0px`. I think I might pad it out to `2px` which helps a little, tweaked in 088014c.

**Comment by lurch** [comment]
https://github.com/raspberrypi/documentation/pull/2092/files has `margin-bottom: 0px;` but I don't know CSS.

**Comment by aallan** [comment]
The real problem with the Sense HAT stuff is there should be *that many* level 2 headers.

**Comment by aallan** [comment]
The Sense HAT page breaks the paradigm since it's all in the top level file rather than spilt out into separate files. Hmm.

**Comment by aallan** [comment]
Waaaaiiit… those are level 2 headers. Why don't they have a "Edit this on Github" below them @lurch ?

**Comment by lurch** [comment]
It's not really a "paradigm", more "convention for very long pages".

**Comment by lurch** [comment]
It only adds a "Edit" link for the _first_ header it finds on an included (or top-level in the case of Sense-HAT) page.

**Comment by aallan** [comment]
Right. Soooo… that sounds like we need to somehow tag H2 headers that won't be getting an "Edit this" link below them with something @nelliemckesson can hang some CSS off the tag so it increases that `bottom-margin` back up again so there's some space below them.

So effectively we have different CSS for H2 headers that have edit links, and those that do not?

**Comment by aallan** [comment]
Okay. good suggestion on chat there, flipping things around making the H2 header have a `bottom-margin: 15px` and the edit-this div style have a `top-margin: -12px` seems to sort this out. See d89a4b5.

**Comment by aallan** [comment]
I think there is a *bit* more space between the "Edit this on Github" link and the top of the opening para now, but it still looks okay I think?

**Comment by aallan** [comment]
Pushed #2093.

**Comment by aallan** [comment]
Oh the CSS is cached! 🤦‍♂️ 

That's be 30 min or an hour before it looks right then!

**Comment by lurch** [comment]
I think @mudge mentioned he has a way of manually flushing the Cloudflare cache?

**Comment by mudge** [comment]
I've purged the Cloudflare cache for https://www.raspberrypi.org/documentation/css/style.css so you should see the latest version now.

**Pull request merged**